internal/controller/managedcluster_controller.go

```diff
@@ -537,32 +518,49 @@ func (r *ManagedClusterReconciler) Delete(ctx context.Context, managedCluster *h
 }
 
 func (r *ManagedClusterReconciler) releaseCluster(ctx context.Context, namespace, name, templateName string) error {
-	providers, err := r.getProviders(ctx, namespace, templateName)
+	providers, err := r.getInfraProvidersNames(ctx, namespace, templateName)
 	if err != nil {
 		return err
 	}
 
-	for _, provider := range providers.BootstrapProviders {
-		if provider.Name == "eks" {
-			// no need to do anything for EKS clusters
-			return nil
+	var (
+		gvkAWSCluster = schema.GroupVersionKind{
+			Group:   "infrastructure.cluster.x-k8s.io",
+			Version: "v1beta2",
+			Kind:    "AWSCluster",
+		}
+
+		gvkAzureCluster = schema.GroupVersionKind{
+			Group:   "infrastructure.cluster.x-k8s.io",
+			Version: "v1beta1",
+			Kind:    "AzureCluster",
 		}
-	}
+
+		gvkMachine = schema.GroupVersionKind{
+			Group:   "cluster.x-k8s.io",
+			Version: "v1beta1",
+			Kind:    "Machine",
+		}
+	)
 
 	providerGVKs := map[string]schema.GroupVersionKind{
 		"aws":   gvkAWSCluster,
 		"azure": gvkAzureCluster,
 	}
 
 	// Associate the provider with it's GVK
-	for _, provider := range providers.InfrastructureProviders {
-		gvk, ok := providerGVKs[provider.Name]
+	for _, provider := range providers {
+		gvk, ok := providerGVKs[provider]
 		if !ok {
 			continue
 		}
 
 		cluster, err := r.getCluster(ctx, namespace, name, gvk)
 		if err != nil {
+			if provider == "aws" && apierrors.IsNotFound(err) {
```

Comment:
Could you please explain this condition to me? The `providers` array collects infra providers if it has `infrastructure-` prefix. So, the `provider` value will never be `aws`. Also, I don't understand why this condition is needed but I'm probably missing something.

Reply:
That's a mistake putting names without cutting the prefixes, thanks for the note! Fixed. Regarding the error, there was an issue fixed within #492 , and since we have only the `aws` provider (the `eks` one was made up) we have to skip errors when the `AWSCluster` object is missing (there are no such objects in case of the `eks` deployment and there is no way to distinguish one from another)